File: dump-cn/930.binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] A, int S) {
        int m = A.length;
        HashMap<Integer, Integer> dic = new HashMap<Integer, Integer>();
        int ptr = 0;
        int count = 0;
        int sum = 0;
        dic.put(0,1);
        while (ptr!=m)
        {
            sum += A[ptr];
            count += dic.getOrDefault(sum - S, 0);
            dic.put(sum, dic.getOrDefault(sum, 0) + 1);
            ptr++;
        }
        return count;
    }
}

File: dump-cn/930.binary-subarrays-with-sum.cs
public class Solution {
    public int NumSubarraysWithSum(int[] A, int S)
    {
        int m = A.Length;
        Dictionary<int, int> dic = new Dictionary<int, int>();
        int ptr = 0;
        int count = 0;
        int sum = 0;
        dic[0] = 1;
        while (ptr!=m)
        {
            sum += A[ptr];
            if (dic.ContainsKey(sum - S))
                count += dic[sum - S];
            if (dic.ContainsKey(sum))
                dic[sum]++;
            else
                dic[sum] = 1;
            ptr++;
        }
        return count;
    }
}